core/core.js

```diff
@@ -1,411 +1,123 @@
-  ////////////////////////////////////// Async handling //////////////////////////////////////
-  ////////////////////////////////////////////////////////////////////////////////////////////
-
-  let asyncHandlers = [];
-
-  function setAsyncHandler(opId, cb) {
-    assert(opId != null);
-    asyncHandlers[opId] = cb;
+  function getErrorClassAndArgs(errorName) {
+    return errorMap[errorName] ?? [undefined, []];
   }
 
-  function handleAsyncMsgFromRust() {
-    while (true) {
-      const opIdBuf = shift();
-      if (opIdBuf == null) {
-        break;
-      }
-      assert(asyncHandlers[opIdBuf[0]] != null);
-      asyncHandlers[opIdBuf[0]](opIdBuf[1], true);
-    }
+  let nextPromiseId = 1;
+  const promiseTable = new Map();
 
-    for (let i = 0; i < arguments.length; i += 2) {
-      asyncHandlers[arguments[i]](arguments[i + 1], false);
+  function processResponse(res) {
+    // const [ok, err] = res;
+    if (res[1] === null) {
+      return res[0];
     }
+    throw processErr(res[1]);
   }
 
-  ////////////////////////////////////////////////////////////////////////////////////////////
-  ///////////////////////////// General sync & async ops handling ////////////////////////////
-  ////////////////////////////////////////////////////////////////////////////////////////////
-
-  let nextRequestId = 1;
-  const promiseTable = {};
-
-  function asyncHandle(u8Array, isCopyNeeded, opResultParser) {
-    const [requestId, result, error] = opResultParser(u8Array, isCopyNeeded);
-    if (error !== null) {
-      promiseTable[requestId][1](error);
-    } else {
-      promiseTable[requestId][0](result);
+  function processErr(err) {
+    const [ErrorClass, args] = getErrorClassAndArgs(err.className);
+    if (!ErrorClass) {
+      return new Error(
+        `Unregistered error class: "${err.className}"\n  ${err.message}\n  Classes of errors returned from ops should be registered via Deno.core.registerErrorClass().`,
+      );
     }
-    delete promiseTable[requestId];
+    return new ErrorClass(err.message, ...args);
   }
 
-  function opAsync(opName, opRequestBuilder, opResultParser) {
-    const opId = opsCache[opName];
-    // Make sure requests of this type are handled by the asyncHandler
-    // The asyncHandler's role is to call the "promiseTable[requestId]" function
-    if (typeof asyncHandlers[opId] === "undefined") {
-      asyncHandlers[opId] = (buffer, isCopyNeeded) =>
-        asyncHandle(buffer, isCopyNeeded, opResultParser);
-    }
-
-    const requestId = nextRequestId++;
-
-    // Create and store promise
-    const promise = new Promise((resolve, reject) => {
-      promiseTable[requestId] = [resolve, reject];
+  async function jsonOpAsync(opName, args = null, zeroCopy = null) {
+    const promiseId = nextPromiseId++;
+    const maybeError = dispatch(opName, promiseId, args, zeroCopy);
+    // Handle sync error (e.g: error parsing args)
+    if (maybeError) processResponse(maybeError);
```

Comment:
Wrap in braces

Reply:
I had wrapped this in braces and then `./tools/format.js` undid that.